File: Algorithms/132_PalindromePartitioningII.cpp
class Solution {
public:
    // http://fisherlei.blogspot.jp/2013/03/leetcode-palindrome-partitioning-ii.html
    // dp
    int minCut(string s) {
        const int len = s.size();
        int f[len+1];
        bool p[len][len];
        // the worst case is cutting by each char
        for (int i = 0; i <= len; i++)
            f[i] = len - 1 - i; // 最后一个 f[len]=-1
        for (int i = 0; i < len; i++)
            for (int j = 0; j < len; j++)
                p[i][j] = false;
        for (int i = len - 1; i >= 0; i--) {
            for (int j = i; j < len; j++) {
                if (s[i] == s[j] && (j - i < 2 || p[i + 1][j - 1])) {
                    p[i][j] = true;
                    f[i] = min(f[i], f[j + 1] + 1);
                }
            }
        }
        return f[0];
    }
};

File: Algorithms/132.PalindromePartitioningII.py
class Solution:
    # @param {string} s
    # @return {integer}
    def minCut(self, s):
        n = len(s)
        cuts = [n-i-1 for i in range(n+1)]
        isPalindrome = [[False for i in range(n)] for j in range(n)]
        for i in range(n)[::-1]:
            for j in range(i, n):
                if s[i] == s[j] and (j - i < 2 or isPalindrome[i + 1][j - 1]):
                    isPalindrome[i][j] = True
                    cuts[i] = min(cuts[j+1] + 1, cuts[i])
        return cuts[0]
